### PR TITLE
12549 added Registry Your Business functionality

### PR DESCRIPTION
**Pull request by severinbeauvais**
*Issue #:* bcgov/entity#12549

*Description of changes:*
- added Dashboard URL config key
- deleted obsolete configuration keys
- updated Affiliation Error Dialog
- added Show Register Button condition
- added Register Your Business event handler
- updated NR Approved Gray Box
- added firm entity types (and sorted the enum)
- added Business Request optional object
- added isFirm()
- updated createAffiliation()
- added Dashboard URL key fetch
- updated createNrAffiliation()
- updated createBusiness()
- added fetchBusiness()
- added feature flag
- naviagte to BC Registry instead of using Incorporate Login Modal
- fixed App logged in processing
- removed NR_DATA after affiliation
- misc cleanup
- app version = 3.1.0

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namerequest license (Apache 2.0).

### CODE REVIEW

File: public/config/configuration.json
Comment (severinbeauvais) [inline public/config/configuration.json:7]: This is in `web-url` vault, `business` application, which is already being loaded (see ~/devops/vaults.json).

File: public/config/configuration.json
Comment (severinbeauvais) [inline public/config/configuration.json:7]: These are no longer referenced in the code.

File: src/components/dialogs/affiliation-error.vue
Comment (severinbeauvais) [inline src/components/dialogs/affiliation-error.vue:12]: This text needs to be approved.

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:24]: This format (PascalCase instead of kebab-case) allows you to click F12 on this component and it will go to its implementation.

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:417]: This is trivial atm but it might become more complex later, and this also is a nice abstraction from what the component needs to know (do I show the button?) to what the business logic is (is this a firm?).

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:653]: Again, trivial atm but it might become more complex later.

Also this lets me add in a safety check (so I don't have to assume that the button can only be clicked when the NR is approved or conditional).

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:79]: This modal is displayed even when there is a createBusiness() error, so it displays an invalid message.

I didn't fix this here but I didn't break it more, either :)

Maybe this can be fixed in the future.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:37]: Possible future improvement: if the NR is already affiliated, try to create the business; if the business already exists, go to its dashboard.

Implementing this would require API changes, since it will need to provide the Business Identifier if already affiliated (otherwise there is no way to get it).

Or, we could navigate to the user's Business Registry page, and they could open the business from there. (future)

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:35]: Actually, it's NOT consumed. It would be nice to know whether the NR is already affiliated, or consumed, or expired, or whatever. Then we could give the user some options (or instructions) here.

This is future.

File: src/services/auth.services.ts
Comment (severinbeauvais) [inline src/services/auth.services.ts:21]: Now this method never returns an exception, and the caller can check the response status.

File: src/services/business.services.ts
Comment (severinbeauvais) [inline src/services/business.services.ts:22]: It is not this method's job to clean up for something that happened in the parent. Hence, this is moved to the caller.

**Comment by severinbeauvais** [comment]
Hi reviewers! Could you have a look at this and give me your feedback and expertise?

I am reusing some existing affiliation/navigation functionality -- but I am not sure if that functionality is currently used? And if not, why not?

File: src/components/existing-request/existing-request-display.vue
Comment (severinbeauvais) [inline src/components/existing-request/existing-request-display.vue:187]: Disabled same as other buttons on line 164.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [reply src/mixins/nr-affiliation-mixin.ts:35]: Actually, something needs to be done here now (*): if I am logged in and create a NR then it is automatically affiliated to my account, so this call fails.

What should happen is that the UI moves to the next step (try to create business), and if the business is already created then navigate to open it.

(*) I am looking into what needs to be updated to do this. In the meantime, this does not block this PR -- it still works OK, ie:

image

Comment (cameron-freshworks) [reply src/mixins/nr-affiliation-mixin.ts:35]: Sorry this is pulling on recollection from ages ago..
I believe you have it right, when there is an account present during the NR creation, i believe an event is published from Namex which Auth picks up on and subsequently auto-affiliates. (Or it could be a request directly to Auth, i can't recall)

I don't know the full context of your work but if you need to add another step here and if* the request isn't directly from Name Request UI to Auth, it might require an update in Namex/Auth. 

Sorry, not a 'totally' useful comment just trying to add some context, if it helps 😁

Comment (cameron-freshworks) [reply src/mixins/nr-affiliation-mixin.ts:35]: 🤔 Just at a quick glance, looks like there is a direct link to the auth services, so maybe not so tricky after all.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [inline src/mixins/nr-affiliation-mixin.ts:35]: Actually, it's not necessarily consumed. It would be nice to know whether the NR is already affiliated, or consumed, or expired, or whatever. Then we could give the user some options (or instructions) here.

---

Actually, something needs to be done now (*): if I am logged in and create a NR then it is automatically affiliated to my account, so this call fails.

What should happen is that the UI moves to the next step (try to create business), and if the business is already created then navigate to open it.

(*) I am looking into what needs to be updated to do this. In the meantime, this does not block this PR -- it still works OK, ie:

image

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [reply src/mixins/nr-affiliation-mixin.ts]: Code is updated. This conversation is obsolete.

File: src/components/dialogs/affiliation-error.vue
Comment (severinbeauvais) [reply src/components/dialogs/affiliation-error.vue:12]: Janis is OK with this text and link for now. She will update this text depending on the possible error scenarios.

File: src/components/dialogs/affiliation-error.vue
Comment (severinbeauvais) [inline src/components/dialogs/affiliation-error.vue:2]: Persistent so that user must click OK to close this dialog. Otherwise, the visibility state is not reset and this modal does not re-display on subsequent errors.

File: src/plugins/featureFlags.ts
Comment (severinbeauvais) [inline src/plugins/featureFlags.ts:11]: I added this FF in LaunchDarkly. It is the same flag as in sbc-auth to allow only SP/GP registrations from a NR (else goes to OneStop).

Currently, it is True in Dev and Test, and False in Prod.

When SP/GP is released to the public, this flag will have to be enabled in Prod (for both sbc-auth and namerequest LD projects).

File: src/views/auth/Signin.vue
Comment (severinbeauvais) [inline src/views/auth/Signin.vue:60]: `createAffiliation` doesn't throw errors, so catch is not needed.

File: src/mixins/nr-affiliation-mixin.ts
Comment (severinbeauvais) [reply src/mixins/nr-affiliation-mixin.ts:45]: Code is updated. This conversation is obsolete.

File: src/App.vue
Comment (severinbeauvais) [inline src/App.vue:10]: Per Janis, this is obsolete. We now redirect to BC Registry site to login (or create a new account).

File: src/App.vue
Comment (severinbeauvais) [inline src/App.vue:205]: Moved to new mixin.

File: src/App.vue
Comment (severinbeauvais) [inline src/App.vue:185]: LD update call was previously missing.

File: src/components/new-request/stats.vue
Comment (severinbeauvais) [inline src/components/new-request/stats.vue:59]: `Watch` wasn't used in this component so I deleted its import.

(I happened upon this file while verifying other things.)

**Review by eve-git — APPROVED** [review]

File: src/components/new-request/stats.vue
Comment (ozamani9gh) [reply src/components/new-request/stats.vue:59]: awesome , thanks bud

**Review by ozamani9gh — APPROVED** [review]
Looks Great Sev.  Great work on this

**Review by cameron-freshworks — APPROVED** [review]
Looks great, nice work!

File: src/mixins/load-keycloak-roles-mixin.ts
Comment (pwei1018) [inline src/mixins/load-keycloak-roles-mixin.ts:1]: Do you think this function can move to common components in the future?

**Review by pwei1018 — APPROVED** [review]

File: src/mixins/load-keycloak-roles-mixin.ts
Comment (severinbeauvais) [reply src/mixins/load-keycloak-roles-mixin.ts:1]: Yes. I just followed the existing code style (eg, NR Affiliation Mixin). Personally I don't mind this as it doesn't pollute the namespace of every component with a method they won't use.

Can this be done in the future maybe?

Comment (cameron-freshworks) [reply src/mixins/load-keycloak-roles-mixin.ts:1]: `vue-property-decorator` isn't a package required for Vue 3 (or Vue 2 that integrates with composition api),
but to fully move away from it would require a rewrite of A LOT of class based components 😅 

Just some thoughts!

Comment (severinbeauvais) [reply src/mixins/load-keycloak-roles-mixin.ts:1]: I think Patrick was referring to this new mixin that I created that contains a single method.

Comment (cameron-freshworks) [reply src/mixins/load-keycloak-roles-mixin.ts:1]: Ah of course 😆  Sorry!

**Pull request merged**
